// react-native-custom-modals/src/AnimatedContainer.js
import React from 'react';
import {
  Animated,
  View,
  TouchableWithoutFeedback,
  StyleSheet
} from 'react-native';

class AnimatedContainer extends React.Component {
  state = {
    isVisible: false,
    animated: new Animated.Value(0)
  };

  show = () => {
    this.setState({ isVisible: true }, () => {
      Animated.timing(this.state.animated, {
        toValue: 1,
        duration: 300,
        useNativeDriver: true
      }).start();
    });
  };

  hide = () => {
    Animated.timing(this.state.animated, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true
    }).start(() => {
      this.setState({ isVisible: false }, this.props.onHide);
    });
  };

  render() {
    return (
      <View style={styles.container}>
        {this.state.isVisible &&
          <TouchableWithoutFeedback onPress={this.hide}>
            <Animated.View
              style={[styles.background, { opacity: this.state.animated }]}
            />
          </TouchableWithoutFeedback>}
        <Animated.View
          style={[
            {
              transform: [
                {
                  translateY: this.props.height
                }
              ]
            },
            this.state.isVisible
              ? {
                  backgroundColor: this.props.backgroundColor || 'white',
                  transform: [
                    {
                      translateY: this.state.animated.interpolate({
                        inputRange: [0, 1],
                        outputRange: [this.props.height, 0]
                      })
                    }
                  ]
                }
              : {}
          ]}
        >
          {this.state.isVisible &&
            <View style={{ height: this.props.height }}>
              {this.props.children}
            </View>}
        </Animated.View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  background: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.4)',
    justifyContent: 'flex-end'
  },

  container: {
    flex: 1,
    backgroundColor: 'transparent',
    justifyContent: 'flex-end'
  }
});

export default AnimatedContainer;
